config version file via `pyproject.toml`

neelasha23 opened the pull request:
Modified the non-setup versioner to read version file path from `pyproject.toml` file.

Closes #58

neelasha23 commented:
Have made the changes. While trying on the `kub-dashboard` repo, the last step where the files are pushed is failing due to permission issues. Adding the stack trace here. @edublancas Need help in testing this

```
(/Users/neelashasen/kub-dashboard) neelashasen@192 kub-dashboard % pkgmt version         
Already on 'main'
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)
hint: Pulling without specifying how to reconcile divergent branches is
hint: discouraged. You can squelch this message by running one of the following
hint: commands sometime before your next pull:
hint: 
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
Already up to date.
Current version in setup.py is 0.1dev. Enter release version (Default: 0.1.0): 

CHANGELOG.md content:

# CHANGELOG

## 0.1.0 (2023-06-26)

## REST API

- [Feature] add `DELETE` `/projects/{project_id}` endpoint
- [API Change] Deletes legacy endpoints
- [Feature] Updates deployment endpoint to also take the input files

## `cloud-core`

- [API Change] `AppRunnerJob.get_logs()` now returns a dictionary with docker and apprunner logs (#31)
- [Feature] Raise error if community user wants to deploy more than one project (#45)
- [API Change] Removes `ProjectModel.new`, we're using the model's constructor now
- [API Change] Move `get_engine` and `set_engine` to root module (#61)
- [API Change] Renames `UserModel.cloud` -> `UserModel.type` and sets `UserType.COMMUNITY` as the default
- [API Change] Enums re-organization
- [API Change] Renames workflow step name from `apprunner-service` -> `webservice`
- [Feature] Add `Project.delete()` for soft delete
- [Feature] Makes `BaseOperations` and abstract class
- [Feature] Add `BaseOperations.update()` generic method
- [API Change] Delete `Operations`
- [API Change] Old `BaseOperations.delete()` is now `BaseOperations._hard_delete()`
- [API Change] `BaseOperations.get_all()` now filters by visibility
- [Feature] Adds `UserModel.visibility` to support soft deletion
- [Doc] tutorial for understanding `Operations` objects

## Developer tools

- [Feature] Running tests with `nox`
- [Doc] Documents how to run tests
- [Fix] Consolidate `pyproject.toml`, `tasks.py`, and `conftest.py`
- [Fix] Consolidates mock data creation (#27)
- [Feature] Adds `invoke doc` for building docs
- [Fix] Fixes incompatibilities with postgres
- [Feature] Adds `--postgres` flag for setting up and running tests
- [Feature] Updates API `Dockerfile` to connect to postgres
- [Feature] Allows updating an existing deployment (#36)

Confirm? [y/n]: y
Commiting release version: 0.1.0
Adding new changes to the repository...
On branch main
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   CHANGELOG.md
	modified:   cloud-core/src/cloud_core/__init__.py

Creating new commit release version...
[main 245a08f] cloud_core release 0.1.0
 2 files changed, 7 insertions(+), 7 deletions(-)
Creating tag 0.1.0...
Pushing tags...
Enumerating objects: 18, done.
Counting objects: 100% (18/18), done.
Delta compression using up to 8 threads
Compressing objects: 100% (11/11), done.
Writing objects: 100% (11/11), 1.14 KiB | 1.14 MiB/s, done.
Total 11 (delta 7), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (7/7), completed with 6 local objects.
To https://github.com/ploomber/kub-dashboard.git
 * [new tag]         0.1.0 -> 0.1.0
Creating new section in CHANGELOG...
Commiting dev version: 0.1.1dev
Adding new changes to the repository...
On branch main
Your branch is ahead of 'origin/main' by 3 commits.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   CHANGELOG.md
	modified:   cloud-core/src/cloud_core/__init__.py

Creating new commit release version...
[main 3d387f9] Bumps up cloud_core to version 0.1.1dev
 2 files changed, 3 insertions(+), 1 deletion(-)
Pushing...
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 8 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 609 bytes | 609.00 KiB/s, done.
Total 7 (delta 5), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (5/5), completed with 5 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/main.
remote: error: Changes must be made through a pull request.
To https://github.com/ploomber/kub-dashboard.git
 ! [remote rejected] main -> main (protected branch hook declined)
error: failed to push some refs to 'https://github.com/ploomber/kub-dashboard.git'
Traceback (most recent call last):
  File "/Users/neelashasen/kub-dashboard/bin/pkgmt", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/pkgmt/cli.py", line 103, in version
    versioneer.version(project_root=".", tag=tag, yes=yes, push=push, target=target)
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/pkgmt/versioneer.py", line 218, in version
    versioner.commit_version(
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/pkgmt/versioner/abstractversioner.py", line 161, in commit_version
    call(["git", "push", "--no-verify"])
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/site-packages/pkgmt/versioner/abstractversioner.py", line 23, in call
    return subprocess.run(*args, **kwargs, check=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/neelashasen/kub-dashboard/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['git', 'push', '--no-verify']' returned non-zero exit status 1.
(/Users/neelashasen/kub-dashboard) neelashasen@192 kub-dashboard % 

```

edublancas commented on doc.md:
```diff
@@ -29,7 +29,13 @@ project_root
 [tool.pkgmt]
```
we're now documenting things in the doc/ directory, please move things over there

edublancas commented on doc.md:
````diff
@@ -29,7 +29,13 @@ project_root
 [tool.pkgmt]
 github = "project/repository" 
 ```
-* Projects which do not contain a `setup.py` file. In this case the tool uses the `_version.py` file.
````
I don't see entries to changelog.md please add. this will be a major version bump, since we're removing some functionality (new version: 0.7.0)

edublancas commented on src/pkgmt/versioneer.py:
```diff
@@ -107,17 +113,24 @@ def version(
-    3. Set new stable version in package_name/_version.py
+    1. These projects should contain `version_file` entry in pyproject.toml file.
+       version_file should be the path to the version file.
+    3. Set new stable version in `version_file`.
```
let's move all of these instructions to the documentation. the docstring here should only contain a brief description of what this function does

edublancas commented on src/pkgmt/versioneer.py:
```diff
@@ -128,9 +141,10 @@ def version(
+    cfg = load()
+    version_file = cfg.get("version_file", None)
+    if version_file:
+        versioner = VersionerNonSetup(version_file, project_root=project_root)
```
seems like the difference between VersionerSetup and VersionerNonSetup is that VersionerSetup will automatically lookup the `__init.__py` while VersionerNonSetup won't. 

maybe it's best to unify the logic, say, in a Versioner class? we can extract the logic in VersionerSetup to lookup the `__init__.py`

something like this:

1. if version_file exists, then use it
2. if version_file does not exist, call a function to lookup `__init__.py`
3. then initialize Versioner

neelasha23 replied:
So should I remove the `AbstractVersioner`? and replace it with `Versioner`

edublancas replied:
yes, we won't need it anymore

edublancas commented on src/pkgmt/versioneer.py:
```diff
@@ -139,7 +153,7 @@ def version(
 
     # TODO: make it compatible with VersionerNonSetup
-    if not version_package and changelog_md_exists:
+    if not version_file and changelog_md_exists:
```
let's get rid of this if statement. the line should be executed regardless of which type of project we're using

edublancas commented on src/pkgmt/versioneer.py:
```diff
@@ -178,7 +192,7 @@ def version(
 
         # TODO: make it compatible with VersionerNonSetup
-        if not version_package:
+        if not version_file:
```
let's get rid of this if statement. the line should be executed regardless of which type of project we're using

edublancas reviewed (changes requested):
you're seeing the error when pushing to kub-dashboard because we have branch protections in place.

you can test the new functionality in this repo: https://github.com/ploomber/dummy - you can add some dummy file to replicate the file structure from kub-dashboard

neelasha23 commented:
> you're seeing the error when pushing to kub-dashboard because we have branch protections in place.
> 
> you can test the new functionality in this repo: https://github.com/ploomber/dummy - you can add some dummy file to replicate the file structure from kub-dashboard

Tested on `ploomber-dummy` repo:  https://github.com/ploomber/dummy/commit/94999410c16d532394fa51a3c620e97a66ac2996

@edublancas

tonykploomber commented on src/pkgmt/config.py:
```diff
@@ -31,13 +30,5 @@ def load():
-        raise FileNotFoundError(
-            "Could not load config: expected pyproject.toml or config.yaml"
-        )
+        raise FileNotFoundError("Could not load config: expected pyproject.toml")
```
Do we have test case for this part?

neelasha23 replied:
yes. `test_config::test_missing_file`.
added the error string also

tonykploomber reviewed (commented):
@neelasha23

Good job! it's the refract of versionnonsetup and versionersetup, it's nicer we can consolidate together now

@edublancas

All looks good to me, except the util.py, might need your help to check:
Any suggestions for `find_package_setup` and `find_package_non_setup`?

We need to verify the correctness of those functions, probably writing some unit test to cover it is one of the ways?

neelasha23 commented on src/pkgmt/versioneer.py:
```diff
@@ -98,26 +105,16 @@ def version(
+        `version_package argument is deprecated, will be removed in version 0.6.3
     """
+
+    if version_package is not None:
+        raise AttributeError(
+            "version_package attribute is deprecated. "
+            "Please add `version_file` in pyproject.toml file."
+        )
+
```
The functionality is same as adding the version_file path in `pyproject.toml`.  If the invoker is not being used elsewhere I think it would be fine to remove it. It was added for the old cloud backend project.
@tonykploomber

edublancas replied:
we can remove this

edublancas commented on doc/versioner.md:
```diff
@@ -0,0 +1,91 @@
+# Versioning a project
+
+Executing `pkgmt version` will modify the `__version__` string in the version file, and the headers in `CHANGELOG.md`
+
```
there's not mention here of how to find the version file (should mention that you put it in a pyproject.toml or in a setup.py by default)

edublancas commented on doc/versioner.md:
```diff
@@ -0,0 +1,91 @@
+|
+|__src
+   |___dir1
+       |___  __init__.py
```
show an example of how the `__init__.py` looks like

edublancas commented on doc/versioner.md:
````diff
@@ -0,0 +1,91 @@
+```
+[tool.pkgmt]
+github = "project/repository" 
+version_file = "/path/to/version_file.py"
````
one last thing, let's modify this so the format is:

```toml
[tool.pkgmt]
github = "project/repository" 
version = {version_file: "/path/to/version_file.py"}
```

this way, we can add new keys to the dictionary to customize the behavior of the `pkgmt version` command (like passing the push or tag arguments)

edublancas commented on src/pkgmt/versioner/util.py:
```diff
@@ -32,3 +36,52 @@ def complete_version_string(version):
+    return package_name, PACKAGE
+
+
+def find_package_non_setup(version_file_path, project_root="."):
```
add docstrings to this function, unclear what's its purpose

edublancas commented on src/pkgmt/versioner/util.py:
```diff
@@ -32,3 +36,52 @@ def complete_version_string(version):
+    return package_name, PACKAGE
+
+
+def find_package_and_version_file(project_root="."):
```
add docstrings to this function, unclear what's its purpose

neelasha23 commented on src/pkgmt/config.py:
```diff
@@ -31,13 +30,5 @@ def load():
-        raise FileNotFoundError(
-            "Could not load config: expected pyproject.toml or config.yaml"
-        )
+        raise FileNotFoundError("Could not load config: expected pyproject.toml")
```
`pyproject.toml` will contain the github repo name, this is being to expand github issues in `changelog.py`. So I think it needs to be present?
@edublancas

edublancas replied:
ah, you're right!

edublancas reviewed (changes requested):
another detail I found, we should raise an exception if the version file does not exist. I got a cryptic error:

```
TypeError: unsupported operand type(s) for /: 'NoneType' and 'str'
```

We should raise a ClickException instead.

neelasha23 commented on doc/versioner.md:
````diff
@@ -0,0 +1,91 @@
+Checks out a tag and uploads to PyPI.
+
+```
+pkgmt release --tag <tag> --production <True/False>
````
`production`  is True by default. How do pass False?
@edublancas

edublancas replied:
if you don't pass the flag, it's set to false

neelasha23 commented on src/pkgmt/versioner/util.py:
```diff
@@ -32,3 +36,52 @@ def complete_version_string(version):
+    else:
+        version_file_name = "__init__.py"
+        package_name, PACKAGE = find_package_setup(project_root)
+    return package_name, PACKAGE, version_file_name
```
Added some specific errors and corresponding test cases:
1. Version file not present
2. Invalid version file path
3. Empty version file path
4. `src` directory doesn't exist

Do you see any other cases where `package_name` will not be found? 

@tonykploomber

tonykploomber replied:
Yes makes sense now, I saw the validation part will be inside of `find_package_of_version_file` and `find_package_in_src`

:+1:

neelasha23 commented:
> another detail I found, we should raise an exception if the version file does not exist. I got a cryptic error:
> 
> ```
> TypeError: unsupported operand type(s) for /: 'NoneType' and 'str'
> ```
> 
> We should raise a ClickException instead.

Added this exception. Resolved the review comments except [this](https://github.com/ploomber/pkgmt/pull/60#discussion_r1258078172).

Please re-review
@edublancas

tonykploomber commented:
@edublancas Please see if this is good to ship

edublancas reviewed (changes requested):
did I do something wrong? doesn't look like this is working (there should be unit tests for these cases)


## test 1: version file does not exist

tried with ploomber/dummy, commit 07ab24107be364a68e1092b1f84f9beff0513bba

```
src
├── ploomber_dummy
│   ├── __init__.py
│   └── __pycache__
│       └── __init__.cpython-39.pyc
└── ploomber_dummy.egg-info
    ├── PKG-INFO
    ├── SOURCES.txt
    ├── dependency_links.txt
    ├── requires.txt
    └── top_level.txt

3 directories, 7 files
```

```
Already on 'main'
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)
Already up to date.
Error: Found the following errors in the project:
- [Inconsistent version] Version in  top section in CHANGELOG is '0.1.41dev', but version in __init__.py is '0.1.40dev', fix them so they match
```

```
[tool.pkgmt]
github = "ploomber/dummy"
version_file = "dummy_another/__init__.py"

```

Expected: 

```
dummy_another/__init__.py does not exist
```

I think it's because we changed the format to:

```
[tool.pkgmt]
github = "ploomber/dummy"
version = {version_file="/path/to/version_file.py"}
```

but then, there should be a warning saying tht "version_file" isn't a valid key (so I can fix it and change it to `version = {version_file="/path/to/version_file.py"}`)


## test two: version_file exists

```
(pkgmt) eduardo@macbookair dev/dummy (main) » pkgmt version                                                                                                                      1 ↵
Already on 'main'
Your branch is ahead of 'origin/main' by 10 commits.
  (use "git push" to publish your local commits)
Already up to date.
Error: Cannot find version file : src/ploomber_another/__init__.py
(pkgmt) eduardo@macbookair dev/dummy (main) » cat src/ploomber_another/__init__.py                                                                                               1 ↵
__version__ = "0.1.44dev"
(pkgmt) eduardo@macbookair dev/dummy (main) »
```

it says the file doesn't exist, but it does!

neelasha23 commented:
Test one:
I think this happened because the versioner couldn't find the `version` key, so it tried to look for `__init__.py` file inside `src` folder. And the `__init__.py` inside `ploomber_dummy` had an outdated version string (because I had tested with `ploomber_another` package earlier so the versions in CHANGELOG.md and the other __init__.py became inconsistent). 

@edublancas

edublancas commented:
ok, so let me open a new issue so we validate the top-keys (test one)

what about `test two: version_file exists`?

neelasha23 commented:
Test two: this was a bug that got introduced after I added the version file path validation. fixed it now : https://github.com/ploomber/dummy/commit/2472918fafe6635b11646224bec6a7b6c9d3cc76
@edublancas
